### ISSUE TITLE
backward() in Autograd Index Function is broken when indexing with LongTensor

### ISSUE THREAD

**Issue by BarclayII** [pytorch/pytorch#828]
In `torch/autograd/_functions/tensor.py:20`:
```python
    def backward(self, grad_output):
        # TODO: this won't have to be zeroed
        grad_input = grad_output.new(self.input_size).zero_()
        grad_input.index(self.index).copy_(grad_output)
        return grad_input
```

I think the index-copy statement doesn't work as intended.
```
In [15]: a = torch.zeros((3, 5))

In [16]: a
Out[16]: 

 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
[torch.FloatTensor of size 3x5]

In [17]: b = torch.ones((2, 5))

In [18]: a.index(torch.LongTensor([0, 2])).copy_(b)
Out[18]: 

 1  1  1  1  1
 1  1  1  1  1
[torch.FloatTensor of size 2x5]

In [19]: a
Out[19]: 

 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
[torch.FloatTensor of size 3x5]
```

I guess the statement should be something like
```python
        grad_input.index_copy_(0, self.index, grad_output)
```

**Comment by soumith** [comment]
you're right. looks like a bug. will fix.

**Comment by apaszke** [comment]
Fixed in #852.
